CloudSQLReplicator/tests/util/cloudsql_test.py

```diff
@@ -0,0 +1,323 @@
+# Copyright 2018 The Pontem Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests GCP API Cloud SQL utility functions."""
+
+import unittest
+import mock
+from mock import MagicMock
+import uuid
+
+import google.auth
+from google.oauth2 import credentials
+
+from google.cloud.pontem.sql.replicator.util import cloudsql
+from google.cloud.pontem.sql.replicator.util import gcp_api_util
+
+TEST_PROJECT_NAME = 'test-project'
+TEST_UUID = uuid.UUID('{12345678-1234-5678-1234-567812345678}')
+TEST_HOST = '127.0.0.1'
+TEST_PORT = '1234'
+TEST_MASTER = 'test-master'
+TEST_DUMPFILE_PATH = 'gs://test_db/employees.sql.gz'
+TEST_REPL_USER = 'repl'
+TEST_PASSWORD = 'password'
+
+@mock.patch.object(
+    google.auth, 'default',
+    return_value=(mock.Mock(spec_set=credentials.Credentials),
+                  TEST_PROJECT_NAME)
+)
+@mock.patch.object(
+    gcp_api_util, 'build_authorized_service'
+)
+class TestSQLAdminMethods(unittest.TestCase):
+    """Test Cloud SQL Admin methods."""
+    def test_build_sql_admin_service(self, mock_build, mock_auth_default):
+        """Test build sql admin service."""
+        _ = cloudsql.build_sql_admin_service()
+        mock_build.assert_called_with(
+            cloudsql.SQL_ADMIN_SERVICE,
+            cloudsql.SQL_ADMIN_SERVICE_VERSION,
+            mock_auth_default.return_value[0]
+        )
+
+    @mock.patch.object(
+        uuid, 'uuid4',
+        return_value=TEST_UUID
+    )
+    def test_create_cloudsql_instance_defaults(self,
+                                               mock_uuid,
+                                               mock_build,
+                                               mock_auth_default):
+        """Test create_cloudsql_instance util function with defaults."""
+        del mock_uuid, mock_auth_default  # unused in body of test
```

Comment:
I'd rather you defined the necessary patches for each test case instead of having class level patches that are not used in some tests. Ditto for other instances of the same.

Reply:
The mocks are used in the tested functions, just not in all of the tests (authorization/web service proxy).